src/plugins/analysis/plugin.py

```diff
@@ -0,0 +1,393 @@
+from __future__ import annotations
+
+import abc
+import ctypes
+import io
+import logging
+import queue
+import re
+import signal
+import time
+import typing
+from multiprocessing import Array, Process, Queue, Value
+
+from pydantic import validator
+from pydantic.dataclasses import dataclass
+
+from objects.file import FileObject
+from storage.fsorganizer import FSOrganizer
+
+# TODO should this be configurable?
+ANALYSIS_STATS_LIMIT = 1000
+
+
+class AnalysisBasePluginAdapterMixin:
+    """A mixin that makes PluginV1 ompatible to AnalysisBasePlugin"""
+
+    # pylint: disable=invalid-name
+
+    def start(self):
+        # This is a no-op
+        pass
+
+    @property
+    def FILE(self):
+        # What is this even used for?
+        raise NotImplementedError
+
+    @property
+    def NAME(self):
+        return self.metadata.name
+
+    @property
+    def DESCRIPTION(self):
+        return self.metadata.description
+
+    @property
+    def DEPENDENCIES(self):
+        return self.metadata.dependencies
+
+    @property
+    def VERSION(self):
+        return self.metadata.version
+
+    @property
+    def RECURSIVE(self):
+        return False
+
+    @property
+    def TIMEOUT(self):
+        raise NotImplementedError
+
+    @property
+    def SYSTEM_VERSION(self):
+        return self.metadata.version
+
+    @property
+    def MIME_BLACKLIST(self):
+        return self.metadata.mime_blacklist
+
+    @property
+    def MIME_WHITELIST(self):
+        return self.metadata.mime_whitelist
+
+    @property
+    def ANALYSIS_STATS_LIMIT(self):
+        # Since no plugin sets this, we just use the default from AnalysisBasePlugin here
+        return ANALYSIS_STATS_LIMIT
+
+    def shutdown(self):
+        # We have no way of knowing which worker to be shut down here
+        # This is a no-op since the PluginRunner already shuts the plugins down
+        pass
+
+
+class PluginV1(metaclass=abc.ABCMeta):
+    """An abstract class that all analysis plugins must inherit from.
+
+    This class MAY NOT depend on FACT_core code.
+    """
+
+    @dataclass
+    class MetaData:
+        """A class containing all metadata that describes the plugin"""
+
+        #: Name of the plugin
+        name: str
+        #: The plugins description.
+        #: MUST be a single sentence.
+        description: str
+        #: The version of the plugin.
+        #: MUST adhere to semantic versioning.
+        version: str
```

Comment:
I think some kind of defined versioning scheme is important and helpful.

Breaking changes would be adaptions to the result schema (which is not implemented yet).
Changes to the view in a backwards compatible way will most probably increase MINOR.
E.g. Fixing a crash of the plugin increases PATCH.

If we can detect breaking changes this way we can show a note to the user that the analysis is outdated and has to be rerun.
Currently this is implemented in the plugin view themselves (e.g. device_tree).